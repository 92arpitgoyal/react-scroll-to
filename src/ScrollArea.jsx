import React, { Component, createRef } from "react";
import PropTypes from "prop-types";
import generateId from "./utilities/generateId";

class ScrollArea extends Component {
  componentDidMount() {
    this.node = createRef();
    this.id = (this.node.curret && this.node.current.id) || generateId();

    this.context.addScrollArea(this.node, this.id);
  }

  componentWillUnmount() {
    this.context.removeScrollArea(this.node.current, this.id);
  }

  render() {
    const { children, ...props } = this.props;

    return (
      <div {...props} ref={this.node}>
        {children}
      </div>
    );
  }
}

ScrollArea.contextTypes = {
  addScrollArea: PropTypes.func.isRequired,
  removeScrollArea: PropTypes.func.isRequired
};

export default ScrollArea;
