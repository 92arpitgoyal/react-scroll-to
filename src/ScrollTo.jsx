import React, { Component, createContext } from "react";
import PropTypes from "prop-types";
import scrollWindow from "./utilities/scrollWindow";

export const ScrollToContext = createContext("scrollToContext");

/**
 * Component that uses render props to inject
 * a function that allows the consumer to scroll to a
 * position in the window or ScrollArea component
 */
class ScrollTo extends Component {
  constructor(props) {
    super(props);

    this.scrollArea = {};

    this.getContext = {
      addScrollArea: (id, ref) => {
        this.scrollArea[id] = ref;
      },
      removeScrollArea: id => {
        delete this.scrollArea[id];
      }
    };
  }

  handleScroll = (props = {}) => {
    const scrollAreaKeys = Object.keys(this.scrollArea);
    const id = props.id || null;
    const x = props.x || 0;
    const y = props.y || 0;

    if (id) {
      this._handleScrollById(id, x, y);
    } else if (scrollAreaKeys.length === 0) {
      scrollWindow(x, y);
    } else {
      scrollAreaKeys.forEach(key => {
        const node = this.scrollArea[key];

        node.scrollLeft = x;
        node.scrollTop = y;
      });
    }
  };

  _handleScrollById = (id, x, y) => {
    const node = this.scrollArea[id];
    if (node) {
      node.scrollLeft = x;
      node.scrollTop = y;
    } else {
      console.warn(`Could not find a ScrollArea with id: ${id}.`);
    }
  };

  render() {
    return (
      <ScrollToContext.Provider value={this.getContext}>
        {this.props.children &&
          this.props.children({
            scrollTo: this.handleScroll
          })}
      </ScrollToContext.Provider>
    );
  }
}

ScrollTo.defaultProps = {
  children: () => {}
};

ScrollTo.propTypes = {
  children: PropTypes.func.isRequired
};

export default ScrollTo;
