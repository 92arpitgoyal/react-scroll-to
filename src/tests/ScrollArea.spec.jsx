import React from "react";
import { render } from "react-testing-library";
import { ScrollToContext } from "../ScrollTo";
import ScrollArea, { ScrollArea as BaseScrollArea } from "../ScrollArea";

jest.mock("../utilities/generateId", () => () => "mock-id");

describe("Test Scroll Area.", () => {
  it("should call addScrollArea when mounting.", () => {
    const addScrollArea = jest.fn();
    const wrapper = render(
      <BaseScrollArea addScrollArea={addScrollArea} removeScrollArea={() => {}}>
        <h1>Test</h1>
      </BaseScrollArea>
    );

    expect(addScrollArea).toHaveBeenCalledTimes(1);
    expect(addScrollArea.mock.calls[0]).toMatchSnapshot();
  });

  it("should call removeScrollArea when unmounting.", () => {
    const removeScrollArea = jest.fn();
    const wrapper = render(
      <BaseScrollArea addScrollArea={() => {}} removeScrollArea={removeScrollArea}>
        <h1>Test</h1>
      </BaseScrollArea>
    );
    wrapper.unmount();

    expect(removeScrollArea).toHaveBeenCalledTimes(1);
    expect(removeScrollArea.mock.calls[0]).toMatchSnapshot();
  });

  it("should render correctly.", () => {
    const wrapper = render(
      <BaseScrollArea
        className="foo"
        addScrollArea={() => {}}
        removeScrollArea={() => {}}
      >
        <h1>Test</h1>
      </BaseScrollArea>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("should render default context.", () => {
    const fns = {
      addScrollArea: () => {},
      removeScrollArea: () => {}
    }

    const wrapper = render(
      <ScrollToContext.Provider value={fns}>
        <ScrollArea style={{ padding: 20 }}>
          test
        </ScrollArea>
      </ScrollToContext.Provider>
    );
  })
});
