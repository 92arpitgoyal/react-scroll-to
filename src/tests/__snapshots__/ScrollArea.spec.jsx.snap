// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Scroll Area. should call addScrollArea when mounting. 1`] = `
Array [
  "mock-id",
  <div>
    <h1>
      Test
    </h1>
  </div>,
]
`;

exports[`Test Scroll Area. should call removeScrollArea when unmounting. 1`] = `
Array [
  "mock-id",
]
`;

exports[`Test Scroll Area. should render correctly. 1`] = `
Object {
  "container": <div>
    <div
      class="foo"
    >
      <h1>
        Test
      </h1>
    </div>
  </div>,
  "getAllByAltText": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getByAltText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "queryAllByAltText": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryByAltText": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
